lightning/src/util/events.rs

```diff
@@ -138,6 +142,10 @@ pub enum Event {
 		/// [`NetworkGraph`]: crate::routing::network_graph::NetworkGraph
 		/// [`NetGraphMsgHandler`]: crate::routing::network_graph::NetGraphMsgHandler
 		network_update: Option<NetworkUpdate>,
+		/// For both single-path and multi-path payments, this is set if all paths of the payment have
+		/// failed. This will be set to false if (1) this is an MPP payment and (2) other parts of the
+		/// larger MPP payment were still in flight when this event was generated.
+		all_paths_failed: bool,
```

Comment:
We really need to rename `PaymentFailed` after this IMO - maybe `PaymentPartFailed` or `HTLCFailed`, because it definitely no longer means the whole payment failed. I'm ok with pushing this to a followup if you just wanna get this merged without blowing up the diff.

Reply:
Agree with renaming `PaymentFailed`. I do like using HTLC in the name as mentioned in another comment. Presumably these won't be deduplicated?

However, I'm still uncertain about adding the `all_paths_failed` field. I feel like I might be missing something, though. How would the retrier make use of it? If false, it needs to retry for the corresponding part unless it has received a `PaymentSent`. If true, doesn't it need to do the same thing? If not, is there any overlap with `rejected_by_dest`?

Reply:
> However, I'm still uncertain about adding the all_paths_failed field. I feel like I might be missing something, though. How would the retrier make use of it? If false, it needs to retry for the corresponding part unless it has received a PaymentSent. If true, doesn't it need to do the same thing?

I believe we'll need a different call in ChannelManager to retry a failed MPP path, as we won't be able to just blindly call send_payment again to keep state consistent. Thus, we'll need to decide which function we call based on that bool.

More generally, the bool isn't *just* for the retrier - its also to tell the user if a *payment* has failed, or if a *path* has failed. Until you either see `all_paths_failed` or `PaymentSent`, you cannot consider a payment resolved.

> If not, is there any overlap with rejected_by_dest?

They're two very different things - whether a final node rejected the payment or not, we can't consider a payment failed unless we get all the paths failed.

Reply:
> I believe we'll need a different call in ChannelManager to retry a failed MPP path, as we won't be able to just blindly call send_payment again to keep state consistent. Thus, we'll need to decide which function we call based on that bool.
> 
> More generally, the bool isn't _just_ for the retrier - its also to tell the user if a _payment_ has failed, or if a _path_ has failed. Until you either see `all_paths_failed` or `PaymentSent`, you cannot consider a payment resolved.
> 

Ah, that's right. For the situation where we retry some paths of an MPP but they all ultimately fail, would the retrier then retry the full payment? Seems it may complicate the retrier as either there are two levels of retries (paths and full payment) or payments that are MPP need to be treated differently from non-MPP. But if we want to avoid the two-levels of retries, we'd also need to know if the payment was MPP. But I don't think the bool will tell us that?

I had hoped this could be generalized in some way where we are always just retrying paths regardless of whether it's MPP or not. But I guess that would require complicating the state tracking in `ChannelManager`, possibly such that the user would need to say when they have given up (and thus considered the payment failed). 🙁

Reply:
Actually, come to think of it the retrier will know how many paths it gets back from the router, so this should be sufficient.

Reply:
> Ah, that's right. For the situation where we retry some paths of an MPP but they all ultimately fail, would the retrier then retry the full payment?

Yes, I believe you generally would, as long as you haven't run out of time/retries.

> Seems it may complicate the retrier as either there are two levels of retries (paths and full payment) or payments that are MPP need to be treated differently from non-MPP. But if we want to avoid the two-levels of retries, we'd also need to know if the payment was MPP. But I don't think the bool will tell us that?

Yes, the complexity is somewhat unavoidable, I think. Ultimately the channelmanager needs to know, when the retry HTLC(s) are sent, whether its an MPP retry or not. If the full MPP has failed, the channelmanager will (have to) forget about the full payment, so the retrier needs to inform the channelmanager of this. The way I was thinking about it was that the retrier would:
 * if `all_paths_failed` is *not* set, call `channelmanager.retry_mpp_part(new_route)` and the channelmanager would look up the pending mpp data, generate the appropriate retry for only the given path, and send the new part(s),
 * if the above call fails with an `Err(NoPendingMPPFound)` or `all_paths_failed` *is* set, the payment retrier will generate a route for the full payment value and call `send_payment`.

I'm not really sure how we could handle this better on the channelmanager end, sadly.

> Actually, come to think of it the retrier will know how many paths it gets back from the router, so this should be sufficient.

Does the retrier need to store the pending route? I figured it would not.

Reply:
> I'm not really sure how we could handle this better on the channelmanager end, sadly.

The only alternative, really, would be to change the API to allow `retry_mpp_part` to reconstruct the state if you provide the full payment value, but I'm not sure that's ideal - if we get a `PaymentFailed` that raced the first `PaymentFailed` and is pending in the queue, with the last node having failed the payment, we *shouldn't* retry, and this API change would result in us retrying, but only part of the MPP.

Reply:
Prefer a quick follow-up for renaming `PaymentFailed`

Reply:
> Does the retrier need to store the pending route? I figured it would not.

Correct. I was thinking it would see that `get_route` returned multiple paths and note that (as a bool) for later. But I guess that isn't necessary if we allow retrying the full payment as well.

For counting retries, were you thinking the number of attempts across all paths and full payments would be counted together? Or would each path get its own attempts? Was thinking the former for simplicity.

As another related concern, what if `get_route` returns multiple paths when attempting to retry a part? Seems like this might be desirable, but `retry_mpp_part` would need to be able to handle this. Maybe that is what you meant by "part(s)".

Reply:
> For counting retries, were you thinking the number of attempts across all paths and full payments would be counted together? Or would each path get its own attempts? Was thinking the former for simplicity.

Dunno, should count something, I'll leave it up to the author :p.

> As another related concern, what if get_route returns multiple paths when attempting to retry a part? Seems like this might be desirable, but retry_mpp_part would need to be able to handle this. Maybe that is what you meant by "part(s)".

Yes, ChannelManager should be able to handle this seamlessly.